amgut/handlers/interactive.py

```diff
@@ -0,0 +1,195 @@
+            taxonomy[5].endswith('__') else highest,
+            'data': value
+        })
+    return datasets
+
+
+class TaxaHandler(BaseHandler):
+    @authenticated
+    def get(self):
+        user = ag_data.get_user_for_kit(self.current_user)
+        bc_info = ag_data.get_barcodes_by_user(user, results=True)
+        titles = [bc['sample_date'].strftime('%b %d, %Y') for bc in bc_info]
+        otus = {}
+        files = []
+        # Load in all possible OTUs that can be seen by loading files to memory
+        # Files are small (5kb) so this should be fine.
+        for barcode in bc_info:
+            with open(join(AMGUT_CONFIG.base_data_dir, 'taxa-summaries',
+                      '%s.txt' % barcode['barcode'])) as f:
+                f.readline()
+                files.append(f.readlines())
+            for line in files[-1]:
+                otus[line.split('\t')[0]] = []
+
+        # Read in counts
+        for file in files:
+            seen_otus = set()
+            for line in file:
+                otu, percent = line.strip().split('\t', 1)
+                otus[otu].append(float(percent) * 100)
+                seen_otus.add(otu)
+            # make the samples without the OTU all zero count
+            unseen = set(otus.keys()) - seen_otus
+            for otu in unseen:
+                otus[otu].append(0.0)
+
+        datasets = _build_taxa(otus)
+
+        meta_cats = ['age-baby', 'age-child', 'age-teen', 'age-20s', 'age-30s',
+                     'age-40s', 'age-50s', 'age-60s', 'age-70+',
+                     'bmi-Underweight', 'bmi-Normal', 'bmi-Overweight',
+                     'bmi-Obese', 'sex-male', 'sex-female', ]
+        self.render('taxa.html', titles=titles, barcodes=bc_info,
+                    meta_cats=meta_cats, datasets=datasets)
+
+
+class MetadataHandler(BaseHandler):
+    @authenticated
+    def get(self):
+        user = ag_data.get_user_for_kit(self.current_user)
+        barcode = self.get_argument('barcode', False)
+        if barcode:
+            barcodes = [b['barcode'] for b in
+                        ag_data.get_barcodes_by_user(user, results=True)]
+            # Make sure barcode passed is owned by the user
+            if barcode not in barcodes:
+                raise HTTPError(403, 'User %s does not have access to barcode '
+                                '%s' % (self.current_user, barcode))
+            site = ag_data.getAGBarcodeDetails(barcode)['site_sampled'].lower()
+        else:
+            site = self.get_argument('site')
+        cat = self.get_argument('category')
+        otus = defaultdict(list)
+        if not cat:
+            file = join(AMGUT_CONFIG.base_data_dir, 'taxa-summaries',
+                        'ag-%s.txt' % site)
+        else:
+            file = join(AMGUT_CONFIG.base_data_dir, 'taxa-summaries',
+                        'ag-%s-%s.txt' % (site, cat))
+
+        with open(file, 'rU') as f:
+            # Read in counts
+            for line in f:
```

Comment:
This is actually a slimmed down version of the original code above. Since it is only caring about one file, this removes a lot of overhead for checking existing otus, etc.